### DISCUSSION TITLE
UI improvements: filling a gap in the Styles workflow

### DISCUSSION THREAD

**Discussion by euu2021** [freeplane/freeplane#546] [Ideas]
I have a general feeling that using Styles is not a fluid experience. I think I identified one rough edge that contributes to that feeling: I think there could be, at the bottom of the format panel, an option to create a new style based on the current properties of the format panel, and immediately apply this new style to the currently selected nodes.

I have in mind this workflow:

- user is editing the map, when he notices that it would be convenient to create a new style for the current node
- user opens the format panel and make some changes, with a live preview of the result in the actual map
- when he is happy with the result, he uses the option at the bottom of the format panel

And also this workflow:

- user is editing the map, when he notices that it would be convenient to change the current node formatting
- user opens the format panel and make some changes
- user likes the result and notices that this could be made into a new style
- user uses the option at the bottom of the format panel

To me, those are very common situations and, with the current UI, there is a bit of friction to accomplish it. 

So, I think this would be one more step to improve the UI. I will create a ticked with that suggestion.

**Comment by euu2021** [comment]
Feature request: #547 

---
One more thing: notice that it would be different from the current `Format->Manage Styles->New user style from selection`, because it goes one step ahead and also applies the style to the current selected node. I really don't see why the user would want to create a new style with the formatting of the current node, and don't apply the new style to that node.

**Reply by dpolivaev** [reply]
Probably, I should change the existing action so that the style is instantly applied.
I think that two actions would be more confusing than the changed one.

**Reply by dpolivaev** [reply]
Implemented in preview 1.10.5_01 (to be uploaded)

**Comment by autoca** [comment]
I have one more thought for improvement:

Sometimes I know that I am going to create a lot of new nodes in a row with a particular style.

It would be nice if there was a setting in the Format panel specifying the style that is applied to any newly created nodes.
Because, if I am going to create a lot of nodes in a row with the same style, I would rather set it once ahead of the whole process instead of having to change the style for each node after having created it.

Unless I overlooked something, right now the "Default" style is applied to new nodes. But I would want a setting that would change the "Default" to another existing style name.
Hope, it is clear enough.

Thx and regards, autoca

**Reply by dpolivaev** [reply]
Nice idea, I am going to implement it.
I am going to add a checkbox "apply current formatting to new nodes" to the Format panel.

**Reply by autoca** [reply]
Wow, big thanks!

**Reply by dpolivaev** [reply]
I see an alternative:

- I could add a check box. If it is checked, the formatting of the currently selected node would be applied.
- I could add a button instead. It would memorize the formatting to apply on all new nodes for a given mind map.

Which suggestion do you like more?

**Reply by autoca** [reply]
Thanks for asking, I had just written a lengthier answer before you posted your latest input. I think it still makes some sense, so I'll post it anyway:

I thought about it some more. Just as a thought to consider:

The format panel shows the attributes of the currently selected node.

So, with a checked checkbox "apply current formatting to new nodes", then you could really only create new nodes with the formatting of the currently selected node (in effect copying the format of the current selection the new - mostly - child, sibling, or new parent). And if I select a different node, than that one's format is copied.

Even though that is interesting, it is not quite what I had in mind. What I had in mind is specifying a particular style for new nodes (the setting of which stays the same until I change it) independently of the style that the currently selected node has. In fact, I don't think it would be of much use (to me) if it would depend on the style of the currently selected node.

But if you think it can be useful to others, please, implement it by all means.
That's what brainstorming is for.

My idea would have been a new drop-down that would hold the style name to be applied for new nodes. Then it would make sense to add a little checkmark next to it, specifying whether or not that setting should be applied, so it can be toggled on/off on the fly.

**Reply by autoca** [reply]
Of course, - if that is not too much - a button that would copy the style name of the current selection to that drop-down can be nice.

(For such a button: In case that the current selection has some formatting that deviates from its actual style, I think that such formatting overrides should be ignored for the new nodes setting. After all, format overrides should only be used in exeptions for individual nodes and a style-for-new-nodes-setting would be meant for creating multiple nodes with the same style.)

**Reply by dpolivaev** [reply]
For me, it would be easier to implement new actions Format->New node style->... similar to Format->Apply style -> ...
I could add an extra value "Use current" to copy the style used by the current selection.

Would it be all right too?

**Reply by autoca** [reply]
You mean in the main menu (only) instead of in the Formats panel?
Sure, if you ask me, as long as it is implemented in some way it is better than not at all. MTIA  -thumbs up-

With such a new feature then, the following question also pops-up:
Should the style name "Default" be left as is? Because it could be confusing to have a style named "Default", both for new users and also for old users who are used to it being applied to new nodes by default, when the new feature actually overrides that behavior. I suggest that its name should at least be changed to "Standard" or similar, because "Default" implies a behavior (to me). And it makes it clear to old users (and remindes them if needed) that there was a change. "Standard", of course, would then be the default setting for "Format->New node style". Of course, I am just brainstorming here.

OT: Does the editor dialog here in the forums support inserting emojis into the text? I can't see it.

**Reply by macmarrum** [reply]
> For me, it would be easier to implement new actions Format->New node style->... similar to Format->Apply style -> ...
I could add an extra value "Use current" to copy the style used by the current selection.

I like the idea :+1: 

Let's keep in mind that Freeplane already has the following functionality:
- `Format->Copy format to new child nodes`
- `Format->Copy format to new sibling nodes`

In my opinion, `Format->New node style->...` should have lower priority than `Copy format to new child nodes` or `Copy format to new sibling nodes`. In other words, if any of the mentioned options is applicable, it is used instead of the new setting.

**Reply by macmarrum** [reply]
> OT: Does the editor dialog here in the forums support inserting emojis into the text? I can't see it.

Nope. An external emoji chooser needs to be used. On Windows, you can use the built-in one: `Win+.`
On Linux, I use [x11-emoji-picker](https://github.com/GaZaTu/x11-emoji-picker)

**Reply by autoca** [reply]
-> emojis: Thank you for the answer.

-> the main topic:
good thinking. I agree to the input about priority.

**Reply by dpolivaev** [reply]
Implemented in preview 1.10.5_01 (to be uploaded)
If both options (set style and copy format) are used but the copied style does not contain a format, the both option are applied.
Otherwise, the "copy format" option has priority.

**Reply by euu2021** [reply]
It is good and useful, but I think the label is not very clear, because it can be read as: set new style for the selected node. 

Maybe something like: `Style of new nodes`

**Reply by dpolivaev** [reply]
Changed to "Style assigned to new nodes". Do you like better?

**Reply by macmarrum** [reply]
I'd rather call it "Style for new nodes".
"Style assigned to new nodes" suggests the assignment was already done.
And the full meaning "Style to be applied to new nodes" is probably too long.

BTW, in other places, the word "apply" is used in conjunction with "style".

**Reply by dpolivaev** [reply]
<img width="148" alt="format_menu" src="https://user-images.githubusercontent.com/1558052/189535405-403b190b-276b-4b66-9a1b-aaa0cd75d448.png">

You are right, it lacks consistency.

What do you think about the following set of changes?

Apply style -> Style
Set new node style -> Preset style for new nodes
Copy format to new sibling nodes -> Preset format for new sibling nodes
Copy format to new child nodes -> Preset format for new child nodes

**Reply by macmarrum** [reply]
If "Apply style" should change, so should "Apply level styles". And if it's changed in menu, it should be changed also in Styles Editor and Format panel.

Since "Apply style" is changed from a verb-sentence to a noun, the same form makes sense for the new action: "Style for new nodes".

As for "Copy format to new sibling nodes", the "preset" version sounds a bit off. The style being applied is dynamic: it depends on the current selection. And _to preset_ is _to know in advance_, which is not the case here. _Copy_ sounds much better to me.

| Old | Dimitry | Mac |
|-|-|-|
| Apply style | Style | Style |
| Apply level styles | | Automatic level styles |
| Set new node style | Preset style for new nodes | Style for new nodes |
| Copy format to new sibling nodes | Preset format for new sibling nodes | Copy format to new sibling nodes |
| Copy format to new child nodes | Preset format for new child nodes | Copy format to new sibling nodes |

**Reply by ldmpub** [reply]
@macmarrum, do you mean for last line and right cell `Copy format for new child nodes`?

**Reply by macmarrum** [reply]
A good catch :+1: 

Yes, it was a copy/paste error

| Old | Dimitry | Mac |
|-|-|-|
| Apply style | Style | Style |
| Apply level styles | | Automatic level styles |
| Set new node style | Preset style for new nodes | Style for new nodes |
| Copy format to new sibling nodes | Preset format for new sibling nodes | Copy format to new sibling nodes |
| Copy format to new child nodes | Preset format for new child nodes | Copy format to new child nodes |

**Reply by dpolivaev** [reply]
Done

**Reply by euu2021** [reply]
Good! 

This made me notice that the label `Copy format for new child nodes` is not precise, because what gets copied is the format AND style. To avoid sidetracking, I will open another thread.

**Comment by ldmpub** [comment]
I agree with @macmarrum proposals ( IMHO `preset` sounds too technical).

Clear / simple english source strings will help translators and in fine end-users.
